### DISCUSSION TITLE
SHAP Marginal Contribution

### DISCUSSION THREAD

**Discussion by itsaugat** [shap/shap#3907] [Q&A]
Hi. I wanted to clarify something about obtaining the marginal contribution of a feature on the outcome. 

I am working with tabular data for binary classification problem. I have trained a RandomForest classifier and initialized the shap explainer as: 

`explainer = shap.TreeExplainer(rf_classifier)`

For a given test instance, `query` I want to get the marginal contribution of a feature, `f` which has column index `f_idx`. 

**Method 1: Directly Using SHAP values**

`shap_values = explainer(query)`
`marginal_1 = shap_values.values[:, label][f_idx]` 

where `label` is the class for which I want to obtain the marginal contribution. 

Here, I am directly getting the SHAP values of the feature to get its marginal contribution.

**Method 2: Using SHAP Interactions**

`shap_interaction_values = explainer.shap_interaction_values(query)`
`marginal_2 = shap_interaction_values[f_idx][f_idx][label]`

Here, I am getting the diagonal value for that feature from SHAP interaction values, which reflect the marginal contribution (or main effect) of that feature excluding the interactions with other features.

Ideally, `marginal_1` and `marginal_2` should be same but I am getting different values. I tried with other test instance and for other features as well. Different values for them as well. 

Any reasons why? And which method could be used to get the most appropriate marginal contribution of a feature?

Many thanks in advance.

**Comment by nicogross** [comment]
They are only the same if the variables are independent. The shap interaction values will sum up to the shap values obtained from a linear regression (the exact shap values). You can have a look [here](https://shap.readthedocs.io/en/latest/example_notebooks/tabular_examples/tree_based_models/Basic%20SHAP%20Interaction%20Value%20Example%20in%20XGBoost.html)
